In [1]:
import pandas as pd
from sklearn.datasets import load_digits
digits = load_digits()

In [2]:
dir(digits)

['DESCR', 'data', 'images', 'target', 'target_names']

In [3]:
digits.target_names

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

In [17]:
print(digits.target)
print(len(digits.target))

[0 1 2 ... 8 9 8]
1797


In [13]:
df = pd.DataFrame(digits.data)
df.head()

0    1    2     3     4     5    6    7    8    9   ...   54   55   56  \
0  0.0  0.0  5.0  13.0   9.0   1.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
1  0.0  0.0  0.0  12.0  13.0   5.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
2  0.0  0.0  0.0   4.0  15.0  12.0  0.0  0.0  0.0  0.0  ...  5.0  0.0  0.0   
3  0.0  0.0  7.0  15.0  13.0   1.0  0.0  0.0  0.0  8.0  ...  9.0  0.0  0.0   
4  0.0  0.0  0.0   1.0  11.0   0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   

    57   58    59    60    61   62   63  
0  0.0  6.0  13.0  10.0   0.0  0.0  0.0  
1  0.0  0.0  11.0  16.0  10.0  0.0  0.0  
2  0.0  0.0   3.0  11.0  16.0  9.0  0.0  
3  0.0  7.0  13.0  13.0   9.0  0.0  0.0  
4  0.0  0.0   2.0  16.0   4.0  0.0  0.0  

[5 rows x 64 columns]

In [15]:
df.shape

(1797, 64)

In [18]:
df['target'] = digits.target
df.head()

0    1    2     3     4     5    6    7    8    9  ...   55   56   57  \
0  0.0  0.0  5.0  13.0   9.0   1.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
1  0.0  0.0  0.0  12.0  13.0   5.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
2  0.0  0.0  0.0   4.0  15.0  12.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
3  0.0  0.0  7.0  15.0  13.0   1.0  0.0  0.0  0.0  8.0  ...  0.0  0.0  0.0   
4  0.0  0.0  0.0   1.0  11.0   0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   

    58    59    60    61   62   63  target  
0  6.0  13.0  10.0   0.0  0.0  0.0       0  
1  0.0  11.0  16.0  10.0  0.0  0.0       1  
2  0.0   3.0  11.0  16.0  9.0  0.0       2  
3  7.0  13.0  13.0   9.0  0.0  0.0       3  
4  0.0   2.0  16.0   4.0  0.0  0.0       4  

[5 rows x 65 columns]

In [19]:
digits.target_names

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

In [20]:
df[df.target==3].head()

0    1     2     3     4     5    6    7    8    9  ...   55   56   57  \
3   0.0  0.0   7.0  15.0  13.0   1.0  0.0  0.0  0.0  8.0  ...  0.0  0.0  0.0   
13  0.0  2.0   9.0  15.0  14.0   9.0  3.0  0.0  0.0  4.0  ...  0.0  0.0  2.0   
23  0.0  1.0   8.0  12.0  15.0  14.0  4.0  0.0  0.0  3.0  ...  0.0  0.0  0.0   
45  0.0  0.0   9.0  16.0  13.0   6.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
59  0.0  1.0  10.0  15.0  11.0   1.0  0.0  0.0  0.0  3.0  ...  0.0  0.0  0.0   

      58    59    60    61   62   63  target  
3    7.0  13.0  13.0   9.0  0.0  0.0       3  
13  12.0  12.0  13.0  11.0  0.0  0.0       3  
23  14.0  15.0  11.0   2.0  0.0  0.0       3  
45  11.0  14.0  12.0   8.0  0.0  0.0       3  
59  13.0  16.0  15.0  10.0  0.0  0.0       3  

[5 rows x 65 columns]

In [21]:
from sklearn.model_selection import train_test_split

In [22]:
df.head()

0    1    2     3     4     5    6    7    8    9  ...   55   56   57  \
0  0.0  0.0  5.0  13.0   9.0   1.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
1  0.0  0.0  0.0  12.0  13.0   5.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
2  0.0  0.0  0.0   4.0  15.0  12.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
3  0.0  0.0  7.0  15.0  13.0   1.0  0.0  0.0  0.0  8.0  ...  0.0  0.0  0.0   
4  0.0  0.0  0.0   1.0  11.0   0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   

    58    59    60    61   62   63  target  
0  6.0  13.0  10.0   0.0  0.0  0.0       0  
1  0.0  11.0  16.0  10.0  0.0  0.0       1  
2  0.0   3.0  11.0  16.0  9.0  0.0       2  
3  7.0  13.0  13.0   9.0  0.0  0.0       3  
4  0.0   2.0  16.0   4.0  0.0  0.0       4  

[5 rows x 65 columns]

In [23]:
X = df.drop(['target'], axis='columns')
X.head()

0    1    2     3     4     5    6    7    8    9   ...   54   55   56  \
0  0.0  0.0  5.0  13.0   9.0   1.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
1  0.0  0.0  0.0  12.0  13.0   5.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
2  0.0  0.0  0.0   4.0  15.0  12.0  0.0  0.0  0.0  0.0  ...  5.0  0.0  0.0   
3  0.0  0.0  7.0  15.0  13.0   1.0  0.0  0.0  0.0  8.0  ...  9.0  0.0  0.0   
4  0.0  0.0  0.0   1.0  11.0   0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   

    57   58    59    60    61   62   63  
0  0.0  6.0  13.0  10.0   0.0  0.0  0.0  
1  0.0  0.0  11.0  16.0  10.0  0.0  0.0  
2  0.0  0.0   3.0  11.0  16.0  9.0  0.0  
3  0.0  7.0  13.0  13.0   9.0  0.0  0.0  
4  0.0  0.0   2.0  16.0   4.0  0.0  0.0  

[5 rows x 64 columns]

In [24]:
y = df.target

In [25]:
y.head()

0    0
1    1
2    2
3    3
4    4
Name: target, dtype: int32

In [26]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)

In [27]:
len(X_train)

1437

In [28]:
len(X_test)

360

In [29]:
len(y_train)

1437

In [30]:
len(y_test)

360

In [31]:
X_train.head()

0    1    2     3     4     5     6    7    8    9   ...   54   55  \
1445  0.0  0.0  2.0  12.0   7.0   0.0   0.0  0.0  0.0  0.0  ...  3.0  0.0   
1556  0.0  0.0  1.0  15.0  14.0   1.0   0.0  0.0  0.0  2.0  ...  0.0  0.0   
543   0.0  0.0  4.0  16.0  16.0  16.0  12.0  0.0  0.0  0.0  ...  0.0  0.0   
637   0.0  0.0  3.0  15.0   2.0   0.0   0.0  0.0  0.0  0.0  ...  0.0  0.0   
646   0.0  0.0  7.0  16.0  15.0   4.0   0.0  0.0  0.0  0.0  ...  1.0  0.0   

       56   57    58    59    60   61   62   63  
1445  0.0  0.0   1.0  13.0  15.0  9.0  0.0  0.0  
1556  0.0  0.0   2.0  16.0  11.0  0.0  0.0  0.0  
543   0.0  0.0   3.0  16.0   9.0  0.0  0.0  0.0  
637   0.0  0.0   5.0  16.0   1.0  0.0  0.0  0.0  
646   0.0  0.0  10.0  16.0  14.0  4.0  0.0  0.0  

[5 rows x 64 columns]

In [32]:
y_train.head()

1445    0
1556    1
543     7
637     4
646     0
Name: target, dtype: int32

In [33]:
X_test.head()

0    1    2     3     4     5    6    7    8    9   ...    54   55  \
734   0.0  0.0  0.0   8.0   3.0   0.0  0.0  0.0  0.0  0.0  ...  15.0  6.0   
925   0.0  0.0  6.0  14.0  13.0   3.0  0.0  0.0  0.0  0.0  ...   2.0  0.0   
1010  0.0  0.0  9.0  12.0  12.0  12.0  6.0  0.0  0.0  1.0  ...   8.0  0.0   
521   0.0  0.0  9.0  14.0  16.0  13.0  2.0  0.0  0.0  0.0  ...   0.0  0.0   
904   0.0  0.0  4.0  11.0  12.0  14.0  0.0  0.0  0.0  0.0  ...   4.0  0.0   

       56   57    58    59    60    61    62   63  
734   0.0  0.0   0.0   5.0  14.0  15.0  10.0  0.0  
925   0.0  0.0   4.0  15.0  16.0   9.0   0.0  0.0  
1010  0.0  0.0   9.0  16.0  16.0  15.0   4.0  0.0  
521   0.0  0.0  10.0  16.0   9.0   0.0   0.0  0.0  
904   0.0  0.0   7.0  13.0  16.0  14.0   1.0  0.0  

[5 rows x 64 columns]

In [34]:
y_test.head()

734     6
925     0
1010    5
521     5
904     9
Name: target, dtype: int32

In [35]:
from sklearn.svm import SVC
model = SVC()

In [36]:
model.fit(X_train, y_train)

C:\Users\Pravin\Anaconda3\lib\site-packages\sklearn\svm\base.py:193: FutureWarning: The default value of gamma will change from 'auto' to 'scale' in version 0.22 to account better for unscaled features. Set gamma explicitly to 'auto' or 'scale' to avoid this warning.
  "avoid this warning.", FutureWarning)


SVC(C=1.0, cache_size=200, class_weight=None, coef0=0.0,
    decision_function_shape='ovr', degree=3, gamma='auto_deprecated',
    kernel='rbf', max_iter=-1, probability=False, random_state=None,
    shrinking=True, tol=0.001, verbose=False)

In [37]:
model.score(X_test, y_test)

0.6138888888888889

In [73]:
model_c = SVC(C=1,kernel='linear')
model_c.fit(X_train, y_train)

SVC(C=1, cache_size=200, class_weight=None, coef0=0.0,
    decision_function_shape='ovr', degree=3, gamma='auto_deprecated',
    kernel='linear', max_iter=-1, probability=False, random_state=None,
    shrinking=True, tol=0.001, verbose=False)

In [74]:
model_c.score(X_test, y_test)

0.9777777777777777

In [75]:
X_test.head()

0    1    2     3     4     5    6    7    8    9   ...    54   55  \
734   0.0  0.0  0.0   8.0   3.0   0.0  0.0  0.0  0.0  0.0  ...  15.0  6.0   
925   0.0  0.0  6.0  14.0  13.0   3.0  0.0  0.0  0.0  0.0  ...   2.0  0.0   
1010  0.0  0.0  9.0  12.0  12.0  12.0  6.0  0.0  0.0  1.0  ...   8.0  0.0   
521   0.0  0.0  9.0  14.0  16.0  13.0  2.0  0.0  0.0  0.0  ...   0.0  0.0   
904   0.0  0.0  4.0  11.0  12.0  14.0  0.0  0.0  0.0  0.0  ...   4.0  0.0   

       56   57    58    59    60    61    62   63  
734   0.0  0.0   0.0   5.0  14.0  15.0  10.0  0.0  
925   0.0  0.0   4.0  15.0  16.0   9.0   0.0  0.0  
1010  0.0  0.0   9.0  16.0  16.0  15.0   4.0  0.0  
521   0.0  0.0  10.0  16.0   9.0   0.0   0.0  0.0  
904   0.0  0.0   7.0  13.0  16.0  14.0   1.0  0.0  

[5 rows x 64 columns]

In [76]:
y_test.head()

734     6
925     0
1010    5
521     5
904     9
Name: target, dtype: int32

### Predict 1

In [81]:
X_test[0:1]

0    1    2    3    4    5    6    7    8    9   ...    54   55   56  \
734  0.0  0.0  0.0  8.0  3.0  0.0  0.0  0.0  0.0  0.0  ...  15.0  6.0  0.0   

      57   58   59    60    61    62   63  
734  0.0  0.0  5.0  14.0  15.0  10.0  0.0  

[1 rows x 64 columns]

In [82]:
y_test[0:1]

734    6
Name: target, dtype: int32

#### model has score -> 0.6138888888888889 predicted "X_test[0:1]" as "9" but "6" is correct

In [83]:
model.predict([X_test[0:1]])

array([9])

#### model_c has score -> 0.9777777777777777 predicted "X_test[0:1]" as "6"

In [85]:
model_c.predict(X_test[0:1])

array([6])

In [86]:
X_test[1:2]

0    1    2     3     4    5    6    7    8    9   ...   54   55   56  \
925  0.0  0.0  6.0  14.0  13.0  3.0  0.0  0.0  0.0  0.0  ...  2.0  0.0  0.0   

      57   58    59    60   61   62   63  
925  0.0  4.0  15.0  16.0  9.0  0.0  0.0  

[1 rows x 64 columns]

### Predict 2

In [87]:
y_test[1:2]

925    0
Name: target, dtype: int32

In [88]:
model.predict([X_test[1:2]])

array([9])

In [90]:
model_c.predict(X_test[1:2])

array([0])